mlir/lib/Driver/CompilerDriver.cpp

```diff
@@ -274,79 +286,108 @@ LogicalResult runEnzymePasses(const CompilerOptions &options,
-                if (!options.keepIntermediate) {
-                    return success();
-                }
-                auto res = pipelineTailMarkers.find(pass);
-                if (res != pipelineTailMarkers.end()) {
-                    for (const auto &pn : res->second) {
-                        std::string outFile = fs::path(std::to_string(pipelineIdx++) + "_" + pn)
-                                                  .replace_extension(".mlir");
-                        {
-                            llvm::raw_string_ostream s{outputs[pn]};
-                            print(s);
-                        }
-                        if (failed(dumpToFile(options, outFile, outputs[pn]))) {
-                            return failure();
-                        }
-                    }
+    // For each pipeline-terminating pass, print the IR into the corresponding dump file and
+    // into a diagnostic output buffer. Note that one pass can terminate multiple pipelines.
+    auto afterPassCallback = [&](Pass *pass, Operation *op) {
+        if (!options.keepIntermediate)
+            return;
+        auto res = pipelineTailMarkers.find(pass);
+        if (res != pipelineTailMarkers.end()) {
+            for (const auto &pipelineName : res->second) {
+                {
+                    llvm::raw_string_ostream s{outputs[pipelineName]};
+                    s << *op;
                 }
-                return success();
-            };
-
-            pm.enableIRPrinting(std::unique_ptr<PassManager::IRPrinterConfig>(
-                new CatalystIRPrinterConfig(printHandler)));
+                dumpToFile(options, output.nextPipelineDumpFilename(pipelineName),
+                           outputs[pipelineName]);
+            }
         }
-    }
+    };
+
+    // For each failed pass, print the owner pipeline name into a diagnostic stream.
+    auto afterPassFailedCallback = [&](Pass *pass, Operation *op) {
+        auto res = passPipelineNames.find(pass);
+        assert(res != passPipelineNames.end() && "Unexpected pass");
+        options.diagnosticStream << "While processing pipeline: " << res->second << "\n";
+    };
+
+    // Output pipeline names on failures
+    pm.addInstrumentation(std::unique_ptr<PassInstrumentation>(
+        new CatalystPassInstrumentation(afterPassCallback, afterPassFailedCallback)));
 
+    // Run the lowering pipelines
     if (failed(pm.run(moduleOp))) {
         return failure();
     }
 
+    // For completeness, dump the IR into files which correspond to possible last empty
+    // pipelines.
+    if (options.keepIntermediate) {
+        for (auto pipelineName : danglingPipelines) {
+            {
+                llvm::raw_string_ostream s{outputs[pipelineName]};
+                s << *moduleOp;
+            }
+            dumpToFile(options, output.nextPipelineDumpFilename(pipelineName),
+                       outputs[pipelineName]);
+        }
+    }
+
```

Comment:
This printing process seems a bit more complicated than needed. Is it possible to do it differently?

What is a dangling pipeline?

Reply:
The dangling pipeline is basically an empty pipeline (the one with no passes in it) in the end of the pipeline list. We handle them here to maintain the invariant "on success, each pipeline must have its dump in the keep-intermediate mode". Without this code, empty trailing pipelines wouldn't be mentioned in dumps.

Reply:
Do we need to support this use case? I feel like it would be best to just have a filtering step in the frontend where we remove the empty pipelines. My reasoning is this:

1. pipelines start with MLIR and end with MLIR.
2. If you need to print the output of the end of a dangling pipeline, then just print the input, which clearly we already have (the `.mlir` field as the base start or the end of the last pipeline in the case you describe above).

Reply:
What I concretely mean is adding the invariant that empty pipelines are effectively no-op compiler transformations including the printing of the IR.

Reply:
I think that current design would result in a simpler user API regarding the dumping and getting pipeline output. The PR version allows users to get the correct output for any pipeline. You are effectively suggesting to change the API and ask users to check if their pipelines are empty or not by themselves. I am not strictly against it, the former was my default preference.

Reply:
I see it in this way: the current suggestion allows users to rely on the fact that whenever they enable `keep_intermediate` they would see dumps for all pipelines they configure no matter what. Same for `get_output`. Pipeline contents might become empty as a result of some user-side automation, so I think it is better to maintain 1-to-1 correspondence between pipelines and results.
Your suggestion involving frontend-side checks may result in some outputs missing (trailing empty pipelines). Not a big deal, and I agree that if we implement it, the source code will look simpler on our side, so I am not entirely sure. But I think I would prefer the API simplicity in this case.
I am ready to implement your suggestion if you are not convinced by now.

Reply:
I am against dangling pipelines. I would strongly prefer their removal :sweat_smile: Thanks!

Reply:
Done